python/src/trezorlib/tools.py

```diff
@@ -310,6 +312,75 @@ def wrapped_f(*args: "P.args", **kwargs: "P.kwargs") -> "Union[MT, R]":
     return decorator
 
 
+def _deprecation_retval_helper(value: Any, stacklevel: int = 0) -> Any:
+    stack = inspect.stack()
+    func_name = stack[stacklevel + 1].function
+
+    warning_text = (
+        f"The return value {value!r} of function {func_name}() "
+        "is deprecated and it will be removed in a future version."
+    )
+
+    # start with warnings disabled, otherwise we emit a lot of warnings while still
+    # constructing the deprecation warnings helper
+    warning_enabled = False
+
+    def deprecation_warning_wrapper(orig_value: Callable[P, R]) -> Callable[P, R]:
+        def emit(*args: P.args, **kwargs: P.kwargs) -> R:
+            nonlocal warning_enabled
+
+            if warning_enabled:
+                warnings.warn(warning_text, DeprecationWarning, stacklevel=2)
+                # only warn once per use
+                warning_enabled = False
+            return orig_value(*args, **kwargs)
+
+        return emit
+
+    # Deprecation wrapper class.
+    # Defined as empty at start.
+    class Deprecated(value.__class__):
+        pass
+
+    # Here we install the deprecation_warning_wrapper for all dunder methods.
+    # This implicitly includes __getattribute__, which causes all non-dunder attribute
+    # accesses to also raise the warning.
+    for key in dir(value.__class__):
+        if not key.startswith("__"):
+            # skip non-dunder methods
+            continue
+        if key in ("__new__", "__init__", "__class__"):
+            # skip some problematic items
+            continue
+        orig_value = getattr(value.__class__, key)
+        if not callable(orig_value):
+            # skip non-functions
+            continue
+        # replace the method with a wrapper that emits a warning
+        setattr(Deprecated, key, deprecation_warning_wrapper(orig_value))
+
+    # construct an instance:
+    if isinstance(value, str):
+        # - for str, do it naively:
+        ret = Deprecated(value)
+        # (this branch should cover all builtin types, but we don't use the wrapper
+        # for anything other than str)
+    else:
+        # - for other types (we assume it's MessageType so a user-defined class)
```

Comment:
nit: consider adding an explicit type check?

Reply:
reworked in f99c9f46b044